### PR TITLE
New task structure

### PR DESCRIPTION
**Pull request by osanseviero**
**First part of 4 for #83** 
Internal change at https://github.com/huggingface/moon-landing/pull/2514

This PR changes how we define the types into a way that is better structured, more flexible, and give us more control (e.g. define if we want to show a type just in models or just in datasets). 

**One user-facing change is the order in which the tags are displayed.** I'm happy to revert to previous impl if that's wanted.

Some TODOs/questions

- [x] Add docstrings
- [x] For `PipelineType`, we can instead define `const ALL_PIPELINE_TYPES = [ "text-classification", ... ] as const;` and then `const PIPELINES: Record<PipelineType, PipelineData> = { ... }`, I left this one as it worked fine and you just have to define the task in a single place. Maybe the other approach was preferred
- [x] Fix color of icons
- [x] Remove `PIPELINE_TAGS_DISPLAY_ORDER` 
- [x] Change imports to don't import type for data
- [x] Remove `[]`s since they are optional

Some screenshots showing unaffected changes

Screenshot from 2022-03-31 14-59-14
Screenshot from 2022-03-31 14-59-22

### CODE REVIEW

File: js/src/lib/components/InferenceWidget/shared/WidgetHeader/WidgetHeader.svelte [hunk 1]
```diff
@@ -9,7 +9,7 @@
 
 	export let noTitle = false;
 	export let title: string | null = null;
-	export let pipeline: keyof typeof PipelineType | undefined;
+	export let pipeline: keyof typeof PIPELINE_TYPE_DATA | undefined;
```
Comment (julien-c) [inline js/src/lib/components/InferenceWidget/shared/WidgetHeader/WidgetHeader.svelte:12]:
```suggestion
	export let pipeline: PipelineType | undefined;
```

File: js/src/lib/components/InferenceWidget/shared/WidgetHeader/WidgetHeader.svelte [hunk 2]
```diff
@@ -1,5 +1,5 @@
 <script lang="ts">
-	import type { PipelineType } from "../../../../interfaces/Types";
+	import type { PIPELINE_TYPE_DATA } from "../../../../interfaces/Types";
```
Comment (julien-c) [inline js/src/lib/components/InferenceWidget/shared/WidgetHeader/WidgetHeader.svelte:2]:
```suggestion
	import type { PipelineType } from "../../../../interfaces/Types";
```

File: js/src/lib/interfaces/Types.ts [hunk 3]
```diff
@@ -1,64 +1,179 @@
...
-	"object-detection"                                        = "Object Detection",
-	"image-segmentation"                                      = "Image Segmentation",
-	"text-to-image"                                           = "Text-to-Image",
-	"image-to-text"                                           = "Image-to-Text",
-	/// others
-	"structured-data-classification"                          = "Structured Data Classification",
-	"reinforcement-learning"                                  = "Reinforcement Learning",
+export interface PipelineData {
+	name: string;
+	subtasks: SubTask[];
```
Comment (julien-c) [inline js/src/lib/interfaces/Types.ts:8]:
```suggestion
	subtasks?: SubTask[];
```
maybe make it optional to not have to specify an empty array below?

File: js/src/lib/utils/ViewUtils.ts [hunk 4]
```diff
@@ -102,10 +102,10 @@ export function hexToRgb(hex: string): number[]{
 }
 
 // Get the Task id corresponding to the modelPipeline (should be === in 99% cases)
-export function getPipelineTask(modelPipeline: keyof typeof PipelineType): string {
+export function getPipelineTask(modelPipeline: PipelineType): string {
 	return modelPipeline === "text2text-generation"
 		? "text-generation"
-		: modelPipeline;
+		: "hello";
```
Comment (julien-c) [inline js/src/lib/utils/ViewUtils.ts:108]:
```suggestion
		: modelPipeline;
```

File: tasks/src/audio-classification/data.ts [hunk 5]
```diff
@@ -35,7 +35,7 @@ const taskData: TaskData = {
 		],
 	},
 	id:        "audio-classification",
-	label:     PipelineType["audio-classification"],
+	label:     PIPELINE_DATA["audio-classification"]["name"],
```
Comment (julien-c) [inline tasks/src/audio-classification/data.ts:38]:
```suggestion
	label:     PIPELINE_DATA["audio-classification"].name,
```

**Review by julien-c — COMMENTED** [review]
left a few first comments!

**Comment by julien-c** [comment]
@Pierrci @coyotte508 when you have some time (not urgent), can you take a look at the specific question of:

whether we want to type things like in this current PR, or do `const ALL_PIPELINE_TYPES = [ "text-classification", ... ] as const;` and then `const PIPELINES: Record<PipelineType, PipelineData> = { ... },` (which has the advantage of typing the PipelineData subobjects, but the drawback that you need to list each pipeline type twice?

**Comment by coyotte508** [comment]
> @Pierrci @coyotte508 when you have some time (not urgent), can you take a look at the specific question of:
> 
> whether we want to type things like in this current PR, or do `const ALL_PIPELINE_TYPES = [ "text-classification", ... ] as const;` and then `const PIPELINES: Record<PipelineType, PipelineData> = { ... },` (which has the advantage of typing the PipelineData subobjects, but the drawback that you need to list each pipeline type twice?

How it's done here is perfectly fine for me.

But we can have the [best of both worlds](https://www.typescriptlang.org/play?#code/C4TwDgpgBACglpANnAdhAIgQ2JqBeKAbwFgAoKKAZzgC8IAuKFAVwFsAjCAJwG4yBfMmQBmzFAGNgcAPYooEFJWZcIAFXAQAPAGkIISvIAewBQBMDlYF1QBzAHwAKFeOldTjAEoQXbnXsoANLAIEMhoWDh2AJSe3q6mfvpB8EioGNiYdkRkFCrAynLO8XykgqRkLorAUKYZ+PKKymoaDiTkUABEYCFhEACMHYxtFBTUdIx9AAzTATlQ-LPtXT1pAEyD2e2jtAxQU5OLFGX8USVkoJBQuiD1ANZ60sJQFxCPNRk8UAD0X53dqWgBlAAD5-FZodZCUiVSzvHAAMUwcEQ9QUShU6kgrTmywB-Q2wxGY2aYGkE2mBzmCxx-1CawJc224z2MypAlOUKhQA) if you want :)

**Comment by julien-c** [comment]
reminder to self to read @coyotte508's comment on Monday

**Comment by julien-c** [comment]
i have rebased this branch on `main` locally @osanseviero, i can either push --force here, or push to a new branch (or not do anything 😅) let me know if it's helpful

**Comment by julien-c** [comment]
PS/ @coyotte508's typing solution works and is very cool 👍

**Comment by osanseviero** [comment]
Feel free to push to this branch!

**Review by julien-c — APPROVED** [review]
modulo re-adding a `PIPELINE_TAGS_DISPLAY_ORDER` (that we can change to feature more non-NLP stuff) this LGTM! 🚢

**Review by julien-c — APPROVED** [review]
ok this is ready to merge i think!

**Comment by osanseviero** [comment]
Should we wait until moon landing PR is ready before merging?

**Comment by julien-c** [comment]
no i think we can merge this and I'll rebase/finish the moon-landing PR this afternoon 👍

**Pull request merged**